clients/drcachesim/tracer/instru_online.cpp

```diff
@@ -313,20 +313,34 @@ online_instru_t::instrument_memref(void *drcontext, instrlist_t *ilist, instr_t
 int
 online_instru_t::instrument_instr(void *drcontext, void *tag, void **bb_field,
                                   instrlist_t *ilist, instr_t *where, reg_id_t reg_ptr,
-                                  int adjust, instr_t *app)
+                                  int adjust, instr_t *app, bool scatter_gather_expanded,
+                                  bool repstr_expanded)
 {
-    bool repstr_expanded = *bb_field != 0; // Avoid cl warning C4800.
-    app_pc pc = repstr_expanded ? dr_fragment_app_pc(tag) : instr_get_app_pc(app);
+    // We cannot rely on the given instr's app_pc in case of expanded repstr and
+    // scatter/gather sequences. In both cases, this routine is called for the instr
+    // at the top of the expanded sequence (which has its own separate bb).
+    // - in the repstr expansion, this is a jrcxz instr with a fake app pc.
+    // - in the scatter/gather expansion, this is a non-app reg spill instr, which
+    // doesn't have an app_pc set.
```

Comment:
Rather than passing in all these booleans and having special cases here, it seems better to only call this for app instrs.

Reply:
Even if we call this only for app instrs, we'd still need `scatter_gather_expanded` and `repstr_expanded`, to be able to set `type` correctly.

Before this PR, we used `*bb_field != 0` to determine whether this bb is a repstr expansion, which obviated the need to pass in these booleans. But now that we have multiple type of expansions, we need to pass more data in (either as these boolean params, or maybe as a struct or something).

Reply:
Looking: yes, online instru does special-case rep strings to deal with the pc+1 translation.  But it shouldn't.  That special case should disappear with #4865.

Reply:
Reverted the newly added booleans. This PR now keeps the existing logic for repstr expansion.